### PR TITLE
Support Char format for string

### PR DESCRIPTION
**Pull request by fearthecowboy**
Includes:
 - KnownFormat enumeration (simplifies handling of known formats without enforcing types early on)
 - Thread saftey around logger (makes it so tests can run concurrently)
 - example (bug1125) of using AutoRest and Roslyn to compile, load and reflect to verify generation.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@fearthecowboy__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

**Comment by fearthecowboy** [comment]
Fixes #1125 (for c#)

**Comment by tbombach** [comment]
Nice! I'll review this today

**Comment by gitfool** [comment]
I would still prefer a `date-time-offset` format for more flexible use of `DateTimeOffset`.

**Comment by fearthecowboy** [comment]
@gitfool well, the good thing is that I made it significantly easier to add more formats that can be `opt-in` for the code generators. Plus, the tests are much simpler to write.

File: src/modeler/AutoRest.Swagger/Model/SwaggerObject.cs [hunk 1]
```diff
@@ -97,56 +103,66 @@ public ObjectBuilder GetBuilder(SwaggerModeler swaggerSpecBuilder)
...
+        /// This allows languages to not have a specific PrimaryType decided by the Modeler.
+        /// 
+        /// For example, if the Type is DataType.String, and the KnownFormat is 'char' the C# generator 
+        /// will end up creating a char type in the generated code, but other languages will still 
+        /// use string.
+        /// </summary>
+        /// <returns>
+        /// The PrimaryType that best represents this object.
+        /// </returns>
         public PrimaryType ToType()
```
Comment (tbombach) [inline src/modeler/AutoRest.Swagger/Model/SwaggerObject.cs]: This method's logic is so much better now. Awesome!

File: src/generator/AutoRest.CSharp/ClientModelExtensions.cs [hunk 2]
```diff
@@ -403,60 +415,49 @@ public static string CheckNull(string valueReference, string executionBlock)
...
             {
                 string constraintCheck;
-                string constraintValue = constraints[constraint];
+                string constraintValue = (format == KnownFormat.@char) ?$"'{constraints[constraint]}'" : constraints[constraint];
                 switch (constraint)
                 {
                     case Constraint.ExclusiveMaximum:
-                        constraintCheck = string.Format(CultureInfo.InvariantCulture, "{0} >= {1}", valueReference,
-                            constraints[constraint]);
+                        constraintCheck = $"{valueReference} >= {constraintValue}";
```
Comment (tbombach) [inline src/generator/AutoRest.CSharp/ClientModelExtensions.cs]: Love the string interpolation in C# 6!

**Comment by tbombach** [comment]
:+1: LGTM!

File: src/generator/AutoRest.CSharp/ClientModelExtensions.cs (hunk 2 above)
Comment (fearthecowboy) [inline src/generator/AutoRest.CSharp/ClientModelExtensions.cs]: Yeah, me too. It's far less complicated to see, and most of the time it runs faster than format anyway. :D

**Pull request merged**